playwright/fixtures/index.ts

```diff
@@ -35,113 +44,242 @@ export type Fixtures = {
-                    authStatus: "unknown",
-                    lastSyncChange: null,
-                  },
-                }),
-              },
+                  }),
+                },
+                {
+                  name: "slice-machine_isInAppGuideOpen",
+                  value: "false",
+                },
+              ],
+            },
+          ],
+        },
+      });
+
+      // Prevent user to be onboarded if needed
+      if (!onboarded) {
+        context = await browser.newContext({
+          storageState: {
+            cookies: [],
+            origins: [
               {
-                name: "slice-machine_isInAppGuideOpen",
-                value: "false",
+                origin: config.use.baseURL,
+                localStorage: [],
               },
             ],
           },
-        ],
-      },
-    });
+        });
+      }
+
+      // Create new page object with new context
+      const page = await context.newPage();
+
+      // Logout user by default
+      try {
+        await fs.rm(path.join(os.homedir(), ".prismic"));
+      } catch (error) {
+        // Ignore since it means the user is already logged out
+      }
+
+      // Login user if needed
+      if (loggedIn) {
+        // In CI we define a PRISMIC_URL env variable to fasten the tests
+        let prismicUrl: string | undefined = process.env["PRISMIC_URL"];
+
+        // In local we get the Prismic URL from the browser, it helps to avoid
+        // switching manually between Wroom and Prismic
+        if (!prismicUrl) {
+          await page.route(
+            "*/**/_manager",
+            async (route) => {
+              const postDataBuffer = route.request().postDataBuffer() as Buffer;
+              const postData = decode(postDataBuffer) as Record<
+                "procedurePath",
+                unknown[]
+              >;
+
+              if (postData.procedurePath[0] === "getState") {
+                const response = await route.fetch();
+                const existingBody = await response.body();
+                const existingData = (
+                  decode(existingBody) as Record<"data", unknown>
+                ).data as Record<
+                  "env",
+                  {
+                    endpoints: { PrismicWroom: string };
+                  }
+                >;
+
+                // Get Prismic URL from the response of the getState call
+                prismicUrl = existingData.env.endpoints.PrismicWroom;
 
-    const page = await context.newPage();
+                await route.continue();
+              }
+            },
+            // Ensure only the first getState call is intercepted
+            {
+              times: 1,
+            },
+          );
 
-    // Use the fixture value in the test.
-    await use(page);
+          // Visit the page to trigger the getState call and wait for it
+          await page.goto("/");
+          await page.waitForResponse("*/**/_manager");
```

Comment:
Is there a way to check for behavior rather than implementation details? For example, checking that a user is logged in using an element on the page rather than the internal `_manager` endpoint.

Reply:
Here it's the setup and it's also a tricky part. The goal is to get the env URL in order to improve the DX. Being able to test on stage or prod without changing an env variable. For that I want to wait the getState call but not an element in the UI. It's more faster. As I wrote, setup can rely on implementation details because it's shared and only in a unique place.